src/storage/mvcc/txn.rs

```diff
@@ -146,49 +146,63 @@ impl<S: Snapshot> MvccTxn<S> {
-                            });
-                        }
-                    }
-                }
+        if let Some(lock) = self.reader.load_lock(&key)? {
+            // Abort on lock belonging to other transaction.
+            if lock.ts != self.start_ts {
+                return Err(Error::KeyIsLocked {
+                    key: key.into_raw()?,
+                    primary: lock.primary,
+                    ts: lock.ts,
+                    ttl: lock.ttl,
+                });
+            }
+            if lock.lock_type != LockType::Pessimistic {
+                // Duplicated command. No need to overwrite the lock and data.
+                MVCC_DUPLICATE_CMD_COUNTER_VEC.prewrite.inc();
+                return Ok(());
             }
-            // ... or locks at any timestamp.
-            if let Some(lock) = self.reader.load_lock(&key)? {
-                if lock.ts != self.start_ts {
-                    return Err(Error::KeyIsLocked {
+        } else if options.prewrite_pessimistic_lock {
+            // Pessimistic lock does not exist, the transaction should be aborted.
+            info!("prewrite failed (pessimistic lock not found)";
+                    "start_ts" => self.start_ts,
+                    "key" => %key);
+
+            return Err(Error::PessimisticLockNotFound {
+                start_ts: self.start_ts,
+                key: key.into_raw()?,
+            });
+        }
+
+        if !options.skip_constraint_check {
+            if let Some((commit_ts, write)) = self.reader.seek_write(&key, u64::max_value())? {
+                // Abort on writes after our start timestamp ...
+                // If exists a commit version whose commit timestamp is larger than or equal to
+                // current start timestamp, we should abort current prewrite, even if the commit
+                // type is Rollback.
+                //
+                // If the transaction is pessimistic, the data conflict is resolved by `for_update_ts`
+                if commit_ts == self.start_ts
+                    || (options.is_pessimistic_lock.is_empty() && commit_ts > self.start_ts)
+                {
+                    MVCC_CONFLICT_COUNTER.prewrite_write_conflict.inc();
+                    return Err(Error::WriteConflict {
+                        start_ts: self.start_ts,
+                        conflict_start_ts: write.start_ts,
+                        conflict_commit_ts: commit_ts,
                         key: key.into_raw()?,
-                        primary: lock.primary,
-                        ts: lock.ts,
-                        ttl: lock.ttl,
+                        primary: primary.to_vec(),
                     });
                 }
-                // No need to overwrite the lock and data.
-                // If we use single delete, we can't put a key multiple times.
-                MVCC_DUPLICATE_CMD_COUNTER_VEC.prewrite.inc();
-                return Ok(());
+                if should_not_exist {
```

Comment:
Is it possible to use prewrite whose type is INSERT to acquire a pessimistic lock?

Reply:
To acquire a pessimistic lock that should not exist, we use assertion on a pessimistic lock.
https://github.com/ngaut/unistore/pull/173/files#diff-eb428efafdb5d1e15345b9feaaba1e1bR167

Reply:
Yes, it should be checked when acquires pessimistic lock and no need to check when prewrites a pessimistic lock. At the beginning, the implementation doesn't check it, and I want to change it, but I forgot.😂 
I will do it, and add it to future works.

Reply:
BTW, `PessimisticLockRequest` doesn't have `skip_constraint_check` option now. It will always check it.

Reply:
What the difference between `Assertion_NotExist` and `Insert`?

Reply:
Assertion_NotExists is optional in the optimistic transaction, it is used to debug.
But in the pessimistic transaction, the lock type is already `PessimisticLock`, no place for `Insert`, so we use `Assertion_NotExist` instead.

Reply:
If it is a `Insert` mutation, it means the key should not exist. I can check it.

Reply:
I have checked data constraint when acquiring pessimistic locks. I don't check non-pessimistic lock's rollback in a pessimistic transaction when prewrites. Does it matter? @coocood

Reply:
It doesn't matter.